### PR TITLE
add basic nested hyperv on kvm test case

### PR DESCRIPTION
**Pull request by lixuemin2016**
Signed-off-by: lixuemin2016 <xuli@redhat.com>

### CODE REVIEW

**Comment by lixuemin2016** [comment]
Signed-off-by: lixuemin2016 <xuli@redhat.com>

File: qemu/deps/nested_hyperv/hyperv_env.ps1 [hunk 1]
```diff
@@ -0,0 +1,170 @@
+#######################################################################
```
Comment (lixuemin2016) [inline qemu/deps/nested_hyperv/hyperv_env.ps1:1]: Signed-off-by: lixuemin2016 <xuli@redhat.com>

**Comment by peixiu** [comment]
Tested on a host with Haswell CPU, it can be passed.

Tested with Win2019 guest os, commands as:
python3 ConfigTest.py --testcase=nested_hyperv_on_kvm --guestname=Win2019 --clone=yes

 (1/2) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/2) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1387.53 s)
 (2/2) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.nested_hyperv_on_kvm.q35: STARTED
 (2/2) Host_RHEL.m9.u3.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.nested_hyperv_on_kvm.q35: PASS (1053.16 s)

CPU flags as:
-cpu 'Haswell-noTSX',hv_reset,hv_crash,hv-no-nonarch-coresharing=auto,hv_emsr_bitmap,hv_evmcs,+vmx,hv_stimer,hv_synic,hv_vpindex,hv_relaxed,hv_spinlocks=0x1fff,hv_vapic,hv_time,hv_frequencies,hv_runtime,hv_tlbflush,hv_reenlightenment,hv_stimer_direct,hv_ipi,+kvm_pv_unhalt \

**Comment by peixiu** [comment]
@lixuemin2016 For commit log, you can add some description for this fix, It's more helpful for others to understand this patch~

**Comment by lixuemin2016** [comment]
> @lixuemin2016 For commit log, you can add some description for this fix, It's more helpful for others to understand this patch~

@peixiu thank you so much, Peixiu, updated with more descriptions in the commit log.

1. Start L1 Windows VM (w2022,2019,2016,win10)
2. Start L2 BIOS/UEFI VM via booting up fedora image

Test pass for win2022,2019,2016,win10 Gen1/Gen2 VM.  For win11, after enable Hyper-V role in L1, it cannot boot up with cpu flags as below:
-cpu 'Haswell-noTSX',hv_reset,hv_crash,hv-no-nonarch-coresharing=auto,hv_emsr_bitmap,hv_evmcs,+vmx,hv_stimer,hv_synic,hv_vpindex,hv_relaxed,hv_spinlocks=0x1fff,hv_vapic,hv_time,hv_frequencies,hv_runtime,hv_tlbflush,hv_reenlightenment,hv_stimer_direct,hv_ipi,+kvm_pv_unhalt \

**Comment by lixuemin2016** [comment]
close by mistake :)

**Comment by lixuemin2016** [comment]
Test result for w2022 for example:
./ConfigTest.py --testcase=nested_hyperv_on_kvm.q35 --guestname=Win2022 --platform=x86_64 --machines=q35

JOB LOG    : /root/avocado/job-results/job-2023-06-01T14.34-63165d4/job.log
 (1/3) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: STARTED
 (1/3) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1529.38 s)
 (2/3) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.nested_hyperv_on_kvm.q35: STARTED
 (2/3) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.nested_hyperv_on_kvm.q35: PASS (831.57 s)
 (3/3) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.nested_hyperv_on_kvm.q35: STARTED
 (3/3) Host_RHEL.m9.u0.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2022.x86_64.io-github-autotest-qemu.nested_hyperv_on_kvm.q35: PASS (501.29 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2023-06-01T14.34-63165d4/results.html

**Comment by lixuemin2016** [comment]
@YongxueHong Could you please help to review this PR? Thank you so much.

**Comment by YongxueHong** [comment]
Hi, @lixuemin2016 
For the commit signed-off: `Signed-off-by: root <xuli@redhat.com>`, please update the `root` to the your name, thanks.

File: qemu/tests/cfg/nested_hyperv_on_kvm.cfg [hunk 2]
```diff
@@ -0,0 +1,16 @@
+- nested_hyperv_on_kvm:
+    type = nested_hyperv_on_kvm
+    download_url = "https://fedora.mirror.constant.com/fedora/linux/releases/37/Cloud/x86_64/images/Fedora-Cloud-Base-37-1.7.x86_64.qcow2"
```
[the remark this answers is not in the document]
Comment (lixuemin2016) [reply qemu/tests/cfg/nested_hyperv_on_kvm.cfg]: Yes, changed. Thank you for reminder.

File: qemu/tests/nested_hyperv_on_kvm.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
+import time
+import os
+from virttest import error_context
+from avocado.utils import download
+from avocado.utils import process
+from virttest import data_dir
```
[the remark this answers is not in the document]
Comment (lixuemin2016) [reply qemu/tests/nested_hyperv_on_kvm.py]: Yes, changed. Thank you for reminder.

File: qemu/tests/nested_hyperv_on_kvm.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    timeout = int(params.get("login_timeout", 600))
+    session = vm.wait_for_login(timeout=timeout)
+
+    test.log.info("Prepare vhdx file")
+    get_vhdx()
+
+    need_reboot = 0
+    status, output = session.cmd_status_output("powershell Get-VM -ErrorAction SilentlyContinue", timeout=120, safe=True)
```
[the remark this answers is not in the document]
Comment (lixuemin2016) [reply qemu/tests/nested_hyperv_on_kvm.py]: Yes, changed. Thank you for reminder.

File: qemu/tests/cfg/nested_hyperv_on_kvm.cfg [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    image_dir = "/root/avocado/data/avocado-vt/images"
```
[the remark this answers is not in the document]
Comment (lixuemin2016) [reply qemu/tests/cfg/nested_hyperv_on_kvm.cfg]: Yes, changed. Thank you for reminder.

File: qemu/tests/cfg/nested_hyperv_on_kvm.cfg (hunk 2 above)
[the remark this answers is not in the document]
Comment (lixuemin2016) [reply qemu/tests/cfg/nested_hyperv_on_kvm.cfg]: Hi Yongxue,

Thank you so much for your review and reminder, all the suggestions have been updated. 

Best Regards,
Xuemin

**Comment by lixuemin2016** [comment]
@PaulYuuu, @YongxueHong 

Hi Yihuang, Yongxue,

I have updated this PR based on Yongxue's suggestion.

Could you please help to take a look again at this PR when you get time? 
Thank you so much.

Best Regards,
Xuemin

File: qemu/tests/cfg/nested_hyperv_on_kvm.cfg [hunk 6]
```diff
@@ -0,0 +1,15 @@
+- nested_hyperv_on_kvm:
```
Comment (PaulYuuu) [inline qemu/tests/cfg/nested_hyperv_on_kvm.cfg:1]: Consider `only x86_64` and `only Windows`

File: qemu/tests/nested_hyperv_on_kvm.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    timeout = int(params.get("login_timeout", 600))
+    session = vm.wait_for_login(timeout=timeout)
+
+    test.log.info("Prepare vhdx file")
+    get_vhdx()
+
+    need_reboot = 0
+    status, output = session.cmd_status_output("powershell Get-VM \
```
[the remark this answers is not in the document]
Comment (lixuemin2016) [reply qemu/tests/nested_hyperv_on_kvm.py]: Updated, thank you so much for your comment.

File: qemu/tests/nested_hyperv_on_kvm.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        test.log.info("catch error when remove folder, ignore it")
+
+    time.sleep(10)
+
+    # nested_dest=r"C:\nested-hyperv-on-kvm"
+    # copy via deps
+    hyperv_source = os.path.join(data_dir.get_deps_dir(), "nested_hyperv")
+    vm.copy_files_to(hyperv_source, nested_dest)
+
+    # set ExecutionPolicy mainly for winows 11
```
[the remark this answers is not in the document]
Comment (lixuemin2016) [reply qemu/tests/nested_hyperv_on_kvm.py]: Change the description to "set RemoteSigned policy mainly for windows 10/11, it is default for windows server". This powershell command "Set-ExecutionPolicy"  can be executed to windows server (2016/2019/2022), and windows 10/11. 

For windows server 2016/2019/2022, the default policy is 'RemoteSigned', after execute the command, it does not have side effect. For windows 10/11, the default is undefined, so must execute this command.

==============================================================================
Refer to https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_execution_policies?view=powershell-7.3.

RemoteSigned
The default execution policy for Windows server computers.

Undefined
There is no execution policy set in the current scope.
If the execution policy in all scopes is Undefined, the effective execution policy is Restricted for Windows clients and RemoteSigned for Windows Server.

Thank you so much.
Xuemin

File: qemu/tests/cfg/nested_hyperv_on_kvm.cfg (hunk 6 above)
Comment (lixuemin2016) [reply qemu/tests/cfg/nested_hyperv_on_kvm.cfg:1]: Updated, thank you so much for your comment @PaulYuuu

**Review by PaulYuuu — APPROVED** [review]

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Pull request merged**